### PR TITLE
[PF-2053] support `--gcs-bucket` in `terra gcloud builds submit` with default settings

### PR DESCRIPTION
**Pull request by ginaygoogle**
### [PF-2053](https://broadworkbench.atlassian.net/browse/PF-2053)

This PR aims for the ticket is supports the user specify gcs-bucket by `--gcs-bucket` and set the other settings as defaults. like `--async --timeout=2h 
    --gcs-source-staging-dir \$TERRA_blob_data/cloudbuild_source \
    --gcs-log-dir \$TERRA_blob_data/cloudbuild_logs \`

General test
```
ginay-macbookpro:quickstart-docker ginay$ terra gcloud builds submit --gcs-bucket=guangtest2 --tag us-central1-docker.pkg.dev/\$GOOGLE_CLOUD_PROJECT/ml4h/papermill:`date +'%Y%m%d'` 
Nov 03, 2022 1:13:48 AM com.google.auth.oauth2.DefaultCredentialsProvider warnAboutProblematicCredentials
WARNING: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a "quota exceeded" or "API not enabled" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/.
Setting the gcloud project to the workspace project
Updated property [core/project].
Creating temporary tarball archive of 2 file(s) totalling 99 bytes before compression.
Uploading tarball of [.] to [gs://guangtest2bucketname/cloudbuild_source/1667452429.458945-df7fa8c1e7c745c1bfbbbf4f5d46cb7a.tgz]
Created [https://cloudbuild.googleapis.com/v1/projects/terra-vdevel-sunny-lemon-9339/locations/global/builds/f4a7608d-077a-4c25-8783-5644a9850be3].
Logs are available at [ https://console.cloud.google.com/cloud-build/builds/f4a7608d-077a-4c25-8783-5644a9850be3?project=78134895072 ].
ID                                    CREATE_TIME                DURATION  SOURCE                                                                                              IMAGES  STATUS
f4a7608d-077a-4c25-8783-5644a9850be3  2022-11-03T05:13:50+00:00  -         gs://guangtest2bucketname/cloudbuild_source/1667452429.458945-df7fa8c1e7c745c1bfbbbf4f5d46cb7a.tgz  -       QUEUED
Restoring the original gcloud project configuration: terra-vdevel-sunny-lemon-9339
Updated property [core/project].
```

Failure test for put the --gcs-bucket after the gsfile.
```
ginay-macbookpro:quickstart-docker ginay$ terra gcloud builds submit gs://guangtest2bucketname/cloudbuild_source/1667784462.272249-a85c0b38fc584b29a54c39308f86d780.tgz --gcs-bucket=guangtest2
Nov 06, 2022 8:46:57 PM com.google.auth.oauth2.DefaultCredentialsProvider warnAboutProblematicCredentials
WARNING: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a "quota exceeded" or "API not enabled" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/.
Setting the gcloud project to the workspace project
Updated property [core/project].
ERROR: (gcloud.builds.submit) unrecognized arguments: --gcs-bucket=guangtest2 

To search the help text of gcloud commands, run:
  gcloud help -- SEARCH_TERMS
```

Repeat setting test
```
ginay-macbookpro:quickstart-docker ginay$ terra gcloud builds submit --timeout=1 --gcs-bucket=guangtest2
Nov 07, 2022 12:02:45 AM com.google.auth.oauth2.DefaultCredentialsProvider warnAboutProblematicCredentials
WARNING: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a "quota exceeded" or "API not enabled" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/.
Setting the gcloud project to the workspace project
Updated property [core/project].
Creating temporary tarball archive of 2 file(s) totalling 99 bytes before compression.
Uploading tarball of [.] to [gs://guangtest2bucketname/cloudbuild_source/1667797367.12176-72b691eb89b34c7caac5ccaffd857064.tgz]
Created [https://cloudbuild.googleapis.com/v1/projects/terra-vdevel-sunny-lemon-9339/locations/global/builds/a3904876-39b4-49a9-85ac-4a128b0cb605].
Logs are available at [ https://console.cloud.google.com/cloud-build/builds/a3904876-39b4-49a9-85ac-4a128b0cb605?project=78134895072 ].
ID                                    CREATE_TIME                DURATION  SOURCE                                                                                             IMAGES  STATUS
a3904876-39b4-49a9-85ac-4a128b0cb605  2022-11-07T05:02:47+00:00  -         gs://guangtest2bucketname/cloudbuild_source/1667797367.12176-72b691eb89b34c7caac5ccaffd857064.tgz  -       QUEUED
Restoring the original gcloud project configuration: terra-wsm-d-silken-walnut-6109
WARNING: You do not appear to have access to project [terra-wsm-d-silken-walnut-6109] or it does not exist.
Updated property [core/project].
```

### CODE REVIEW

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 1]
```diff
@@ -1,11 +1,33 @@
...
 public class Gcloud extends ToolCommand {
 
+  private static final Logger logger = LoggerFactory.getLogger(Git.class);
+
+  @CommandLine.Option(
+      names = "--gcs-bucket",
+      description = "specify the gcs buket for gcloud builds")
+  public String bucketName;
+
+  @CommandLine.Option(names = "builds", description = "")
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:25]: 25-29 are awkward. Could this be used instead?

https://cs.github.com/DataBiosphere/terra-cli/blob/dfcad196d73e5ffcaafc615e993361d943792020/src/main/java/bio/terra/cli/command/app/Execute.java?q=parameters#L24

You would make `command` public.

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 2]
```diff
@@ -15,4 +37,49 @@ public String getExecutableName() {
   public String getInstallationUrl() {
     return "https://cloud.google.com/sdk/docs/install";
   }
+
+  @Override
+  protected void executeImpl() {
+    workspaceOption.overrideIfSpecified();
+
+    if (bucketName != null && builds && submit) {
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:45]: If GCS bucket isn't set, you should throw an error.

```
if builds submit {
  if GCS bucket isn't set
    throw error
  Now you can assume GCS bucket is set, and proceed.
```
```

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 3]
```diff
@@ -15,4 +37,49 @@ public String getExecutableName() {
...
+      var resource = Context.requireWorkspace().getResource(bucketName);
+      if (Resource.Type.GCS_BUCKET != resource.getResourceType()) {
+        throw new UserActionableException(
+            String.format(
+                "%s %s cannot be cloned because it is not a gcs-bucket",
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:50]: You're not cloning anything. So all `clone` should be removed.

When running `terra gcloud builds submit`, you must pass the resource name of a GCS bucket with `--gcs-bucket`.

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 4]
```diff
@@ -15,4 +37,49 @@ public String getExecutableName() {
...
+                resource.getResourceType(), resource.getName()));
+      }
+      ArrayList<String> cloneCommands =
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:53]: cloneCommands -> commands

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 5]
```diff
@@ -15,4 +37,49 @@ public String getExecutableName() {
...
+                  "builds",
+                  "submit",
+                  "--async",
+                  "--timeout=2h",
+                  "--gcs-source-staging-dir=" + resource.resolve() + "/cloudbuild_source",
+                  "--gcs-log-dir=" + resource.resolve() + "/cloudbuild_logs",
+                  "--tag="
+                      + "us-central1-docker.pkg.dev/$GOOGLE_CLOUD_PROJECT/ml4h/papermill:`date +'%Y%m%d'`"));
+      cloneCommands.addAll(command);
+
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:66]: Log the final `gcloud builds submit` command, so user knows exactly which flags were added

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 6]
```diff
@@ -15,4 +37,49 @@ public String getExecutableName() {
...
+      try {
+        Context.getConfig().getCommandRunnerOption().getRunner().runToolCommand(cloneCommands);
+      } catch (PassthroughException e) {
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:69]: Why catch, why not just let exception propagate? Delete 67, 69-71, 73

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 7]
```diff
@@ -15,4 +37,49 @@ public String getExecutableName() {
...
+          new ArrayList<>(
+              ImmutableList.of(
+                  "gcloud",
+                  "builds",
+                  "submit",
+                  "--async",
+                  "--timeout=2h",
+                  "--gcs-source-staging-dir=" + resource.resolve() + "/cloudbuild_source",
+                  "--gcs-log-dir=" + resource.resolve() + "/cloudbuild_logs",
+                  "--tag="
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:63]: I'm pretty sure Nicole doesn't want this --tag here. I just asked on PF-2053 exactly what she wants.

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 8]
```diff
@@ -1,11 +1,33 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import picocli.CommandLine;
 import picocli.CommandLine.Command;
 
 /** This class corresponds to the second-level "terra gcloud" command. */
 @Command(name = "gcloud", description = "Call gcloud in the Terra workspace.")
 public class Gcloud extends ToolCommand {
 
+  private static final Logger logger = LoggerFactory.getLogger(Git.class);
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:18]: Git -> Gcloud

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+/** Tests for the `terra gcloud builds submit` commands. */
+@Tag("unit")
+public class GcloudBuildsSubmit extends SingleWorkspaceUnit {
+
+  // external bucket to use for creating GCS bucket references in the workspace
+  private BucketInfo externalSharedBucket;
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:27]: externalSharedBucket is never used. Delete.

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    TestCommand.runCommandExpectSuccess("workspace", "set", "--id=" + getUserFacingId());
+    shareUser = TestUser.chooseTestUserWhoIsNot(workspaceCreator);
+    TestCommand.runCommandExpectSuccess(
+        "workspace", "add-user", "--email=" + shareUser.email, "--role=WRITER");
+    shareUser.login();
+    ExternalGCSBuckets.grantReadAccess(
+        externalSharedBucket, Identity.group(Auth.getProxyGroupEmail()));
+
+    // create a dockerfile as building source
+    new File("./Dockerfile").createNewFile();
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:50]: Why is this needed?

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    if (!server.getCloudBuildEnabled()) {
+      return;
+    }
+    workspaceCreator.login();
+
+    // `terra workspace set --id=$id`
+    TestCommand.runCommandExpectSuccess("workspace", "set", "--id=" + getUserFacingId());
+
+    // `terra resource create gcs-bucket --name=$name --format=json`
+    String name = "resourceName";
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:77]: name -> bucketResourceName

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    TestCommand.runCommandExpectSuccess("resource", "create", "gcs-bucket", "--name=" + name);
+
+    // `builds submit --gcs-bucket=bucketName`
+    TestCommand.runCommandExpectSuccess("gcloud", "builds", "submit", "--gcs-bucket=" + name);
+
+    // `terra resource delete --name=$name`
+    TestCommand.Result cmd =
+        TestCommand.runCommand("resource", "delete", "--name=" + name, "--quiet");
+
+    boolean cliTimedOut =
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:87]: Why have this? Command shouldn't time out, right? Delete.

**Review by melissachang — COMMENTED** [review]
Next time, can different tickets go in different PRs? Eg one PR for PF-2053 and one PR for PF-1755.

I assume you ran your test with `-Pserver=verily-devel`.

Can you confirm that if you pass a tag, the tag is used?

**Comment by ginaygoogle** [comment]
> Next time, can different tickets go in different PRs? Eg one PR for PF-2053 and one PR for PF-1755.
> 
> I assume you ran your test with `-Pserver=verily-devel`.
> 
> Can you confirm that if you pass a tag, the tag is used?

Hi Melissa, Because the PF-1755 just one word change, I mean change the "36" to "40". But next time I will do separately. 
Yes. I am sure the `./gradlew runTestsWithTag -PtestTag=unit --tests  "unit.GcloudBuildsSubmit" --i --scan -Pserver=verily-devel` run pass.

File: src/test/java/unit/GcloudBuildsSubmit.java (hunk 12 above)
Comment (ginaygoogle) [reply src/test/java/unit/GcloudBuildsSubmit.java:87]: Hi Melissa. The bucket is currently used by the gcloud building, so resource delete is blocked by waiting for the bucket to be released. During testing, I experienced time out error here, so I expect a succeed or time out here. Note: the resource will finally be guaranteed to clean up since workspace is deleted in the clean up step.

I also tried to set a timeout=1 second for the gcloud building, but the resource delete still get blocked.

Comment (melissachang) [reply src/test/java/unit/GcloudBuildsSubmit.java:87]: Can you point me to the build scan for the failure that you get, when this line is deleted?

File: src/main/resources/servers/verily.json [hunk 13]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "verily",
   "clientCredentialsFile": "verily_secret.json",
+  "cloudBuildEnabled": true,
```
[the remark this answers is not in the document]
Comment (ginaygoogle) [reply src/main/resources/servers/verily.json:4]: Hi Melissa. I found the # could not been used as comment in Json file. I search online https://reqbin.com/json/5wzepqmt/json-comment-example#:~:text=Can%20I%20add%20a%20comment,elements%20will%20still%20be%20data. find should use key and value. 

But unfortunately, it will cause error while list the server https://screenshot.googleplex.com/PQYCTUR9mqp4DyX could not Unrecognized it. 

Maybe we could add the comments into the server.java instead?

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 14]
```diff
@@ -1,11 +1,33 @@
...
+
+  @CommandLine.Option(
```
Comment (yuhuyoyo) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:19]: can we hide this option if cloudbuild is not enabled in the env? 
https://picocli.info/#_model_transformations

File: src/main/java/bio/terra/cli/command/workspace/List.java [hunk 15]
```diff
@@ -69,7 +69,7 @@ private void printText(java.util.List<UFWorkspaceLight> returnValue) {
 
   /** Column information for table output with `terra workspace list` */
   private enum Columns implements ColumnDefinition<UFWorkspaceLight> {
-    ID("ID", w -> w.id, 36, LEFT),
+    ID("ID", w -> w.id, 40, LEFT),
```
Comment (yuhuyoyo) [inline src/main/java/bio/terra/cli/command/workspace/List.java:72]: is this related?  it's just one line. just split it out.

File: src/main/java/bio/terra/cli/command/Main.java [hunk 16]
```diff
@@ -89,9 +89,12 @@ public static int runCommand(String... args) {
     // allow mixing options and parameters for all commands except the pass-through app commands.
     // this is because any options that follow the app command name should NOT be interpreted by the
     // Terra CLI, we want to pass those through to the app instead
+    //
+    // Note: One exception is the gcloud pass-through app. We want to add additional option like
```
Comment (yuhuyoyo) [inline src/main/java/bio/terra/cli/command/Main.java:93]: i don't think we need to add this comment. it's confusing to know what the comment is talking about when the gcloud line is deleted.

**Review by yuhuyoyo — APPROVED** [review]

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 17]
```diff
@@ -1,11 +1,44 @@
...
 
 /** This class corresponds to the second-level "terra gcloud" command. */
-@Command(name = "gcloud", description = "Call gcloud in the Terra workspace.")
+@Command(
+    name = "gcloud",
+    description = "Call gcloud in the Terra workspace.",
+    modelTransformer = Gcloud.SubCmdFilter.class)
 public class Gcloud extends ToolCommand {
 
+  static class SubCmdFilter implements CommandLine.IModelTransformer {
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:21]:
```
// Broad deployments don't have Cloud build enabled yet; don't expose `--gcs-bucket` in that case.
// TODO(PF-2199): After Broad deployments have Cloud build, delete SubCmdFilter/IModelTransformer.

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 18]
```diff
@@ -1,11 +1,44 @@
...
+    public CommandLine.Model.CommandSpec transform(CommandLine.Model.CommandSpec commandSpec) {
+      try {
+        if (!Context.getServer().getCloudBuildEnabled()) {
+          commandSpec.removeSubcommand("--gcs-bucket");
+        }
+      } catch (SystemException e) {
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:27]: Why catch and return? What happens if you replace 23-29 with

```
        if (!Context.getServer().getCloudBuildEnabled()) {
          commandSpec.removeSubcommand("--gcs-bucket");
        }
```

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 19]
```diff
@@ -1,11 +1,44 @@
...
+        return commandSpec;
+      }
+      return commandSpec;
+    }
+  }
+
+  private static final Logger logger = LoggerFactory.getLogger(Gcloud.class);
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:34]: nit usually this is first line in class. Move to line 20.

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 20]
```diff
@@ -15,4 +48,31 @@ public String getExecutableName() {
   public String getInstallationUrl() {
     return "https://cloud.google.com/sdk/docs/install";
   }
+
+  @Override
+  protected void executeImpl() {
+    workspaceOption.overrideIfSpecified();
+    command.add(0, getExecutableName());
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:55]: I assume this can be deleted if you put back `subcommands.get("gcloud").setStopAtPositional(true);` ? Also line 76.

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 21]
```diff
@@ -15,4 +48,31 @@ public String getExecutableName() {
...
+      ArrayList<String> autoCommands =
+          new ArrayList<>(
+              ImmutableList.of(
+                  "--gcs-source-staging-dir=" + resource.resolve() + "/cloudbuild_source",
+                  "--gcs-log-dir=" + resource.resolve() + "/cloudbuild_logs"));
+
+      logger.info("run command: " + command + autoCommands);
+      command.addAll(autoCommands);
+    }
+    // If user doesn't specify gcs-bucket, we still need to add back the digested `builds` and
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:74]: Delete 74-75, no longer relevant

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 22]
```diff
@@ -1,11 +1,44 @@
...
+        return commandSpec;
+      }
+      return commandSpec;
+    }
+  }
+
+  private static final Logger logger = LoggerFactory.getLogger(Gcloud.class);
+
+  @CommandLine.Option(
+      names = "--gcs-bucket",
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:37]: In PF-2053, Nicole suggested using something different from `--gcs-bucket`, since this is resource name, not bucket name. Let's do `--gcs-bucket-resource`.

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 23]
```diff
@@ -1,11 +1,44 @@
...
+      description =
+          "Resource name (not bucket name) of GCS bucket resource. Required for (and only used for) `gcloud builds submit`.")
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:39]: I don't see where you are requiring this flag.

On second thought, let's not require. Then we don't have to change code. :) So this can be

```
"Optional flag only used for `terra gcloud builds submit`. Resource name (not bucket name) of GCS bucket where Cloud Build source and logs will be stored.")
```

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 24]
```diff
@@ -15,4 +48,31 @@ public String getExecutableName() {
...
+  protected void executeImpl() {
+    workspaceOption.overrideIfSpecified();
+    command.add(0, getExecutableName());
+
+    if (gcsBucketResourceName != null) {
+      var resource = Context.requireWorkspace().getResource(gcsBucketResourceName);
+      if (Resource.Type.GCS_BUCKET != resource.getResourceType()) {
+        throw new UserActionableException(
+            String.format(
+                "%s %s cannot builds submit because it is not a gcs-bucket",
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:62]:
```
"--Resource %s was passed into --gcs-bucket-resource, but it is not a GCS bucket. It is %s"
```
```

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java [hunk 25]
```diff
@@ -15,4 +48,31 @@ public String getExecutableName() {
...
+                resource.getResourceType(), resource.getName()));
+      }
+      ArrayList<String> autoCommands =
+          new ArrayList<>(
+              ImmutableList.of(
+                  "--gcs-source-staging-dir=" + resource.resolve() + "/cloudbuild_source",
+                  "--gcs-log-dir=" + resource.resolve() + "/cloudbuild_logs"));
+
+      logger.info("run command: " + command + autoCommands);
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:71]: `run command: ` -> `Final gcloud command: `

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+@Tag("unit")
+public class GcloudBuildsSubmit extends SingleWorkspaceUnit {
+  private TestUser shareUser;
+
+  @BeforeAll
+  @Override
+  protected void setupOnce() throws Exception {
+    super.setupOnce();
+
+    TestCommand.runCommandExpectSuccess("workspace", "set", "--id=" + getUserFacingId());
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:28]: 28-32 aren't needed, right?

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+    shareUser = TestUser.chooseTestUserWhoIsNot(workspaceCreator);
+    TestCommand.runCommandExpectSuccess(
+        "workspace", "add-user", "--email=" + shareUser.email, "--role=WRITER");
+    shareUser.login();
+
+    // create a dockerfile as building source
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:34]:
```
// `gcloud builds submit` requires a Dockerfile in current directory.
```

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+    String bucketResourceName = "resourceName";
+    TestCommand.runCommandExpectSuccess(
+        "resource", "create", "gcs-bucket", "--name=" + bucketResourceName);
+
+    // `builds submit --async --gcs-bucket=bucketName --tag`
+    TestCommand.runCommandExpectSuccess(
+        "gcloud",
+        "builds",
+        "submit",
+        "--async",
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:69]: Delete, this isn't needed right?

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+        "--gcs-bucket=" + bucketResourceName,
+        "--tag=us-central1-docker.pkg.dev/$GOOGLE_CLOUD_PROJECT/ml4h/papermill:`date +'%Y%m%d'`");
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:71]: // `gcloud builds submit` requires --tag flag.
"--tag=mytag");

**Review by melissachang — COMMENTED** [review]
Sorry about the delay

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java (hunk 18 above)
Comment (ginaygoogle) [reply src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:27]: Hi Melissa, I did this before in the first version. However, at the very very beginning, `Context.getServer()` has no server initialized and would throw an error (See line 175 in Context.java). This try catch is to specially handle the corner case at the initialization stage, and only decide to hide the command once we have the enough (server) information.

Comment (melissachang) [reply src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:27]: Thanks. As a general note, for code that's not obvious, it's helpful to add a comment. :)

This is getting complicated. How does this sound:

- Delete lines 21-32
- Print an error if Cloud Build is not enabled and `--gcs-bucket` is set.

That way, if Cloud Build isn't enabled, and user tries to set `--gcs-bucket`, they'll get an error.

Note that noone (no users) is using pools in Broad deployments. It may be years before pools are used.

@yuhuyoyo let me know if you disagree with this suggestion.

File: src/test/java/unit/GcloudBuildsSubmit.java (hunk 26 above)
Comment (ginaygoogle) [reply src/test/java/unit/GcloudBuildsSubmit.java:28]: Hi Melissa. The workspace set is required because the gcs bucket specify need the workspace information

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java (hunk 20 above)
Comment (ginaygoogle) [reply src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:55]: Hi Melissa, this line of code is to add "gcloud" at the beginning of `command` that is executed at line 76. For example, when user input `terra gcloud builds submit ....`, the command get parsed and will stores `["builds", "submit", "...."]` into `command`. Then a "gcloud" string is added at the front so that `["gcloud", "builds", "submit", "...."]` can be executed. The same behavior can be found at Line 48 in Git.java and Line 56 in ToolCommand.java.

File: src/main/java/bio/terra/cli/command/Main.java [hunk 30]
```diff
@@ -91,7 +91,6 @@ public static int runCommand(String... args) {
     // Terra CLI, we want to pass those through to the app instead
     Map<String, CommandLine> subcommands = cmd.getSubcommands();
     subcommands.get("bq").setStopAtPositional(true);
-    subcommands.get("gcloud").setStopAtPositional(true);
```
[the remark this answers is not in the document]
Comment (melissachang) [reply src/main/java/bio/terra/cli/command/Main.java:94]: Discussed offline, this is fine. Good to know `git` does this too.

Can you please document `--gcs-bucket` in README.md?

File: src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java (hunk 18 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/command/app/passthrough/Gcloud.java:27]: Discussed offline. We'll keep this.

File: src/test/java/unit/GcloudBuildsSubmit.java (hunk 26 above)
Comment (melissachang) [reply src/test/java/unit/GcloudBuildsSubmit.java:28]: Discussed offline. Test is setting workspace on line 57. That is what most CLI tests do, they set workspace in beginning.

Lines 28-32 don't look like they're needed. They set up `shareUser` but `shareUser` isn't used by test. Please try deleting.

File: src/test/java/unit/GcloudBuildsSubmit.java [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+
+    // `terra resource delete --name=$name`
```
Comment (melissachang) [inline src/test/java/unit/GcloudBuildsSubmit.java:73]: Lines 73-82 cause test to wait for 5 minutes. To make test not wait for 5 minutes, let's replace 73-82 with:

// Don't delete bucket. Because of PF-2205, CLI will fail and time out after 5 minutes.
// Resource will be deleted when project is deleted.

**Review by melissachang — APPROVED** [review]
